### DISCUSSION TITLE
proto中定义了http.get的req中有对象怎么解析？

### DISCUSSION THREAD

**Discussion by xmsz** [go-kratos/kratos#1398] [Q&A] [answered]
我的结构张这样
```
message GetListReq {
  message GetListReqFilters {
    int32 user_id = 1;
  }

  int32 page = 1;
  int32 page_size = 2;
  GetListReqFilters filters = 3;
}
```
然后请求的时候`http://127.0.0.1:8001/api/v1/list?page=0&page_size=100&filters={"user_id": 1}`

会提示
```
{
    "code": 500,
    "reason": "",
    "message": "parsing field \"filters\": unsupported message type: \"official.service.v1.GetListReq.GetListReqFilters\"",
    "metadata": {}
}
```

遇到这种情况这么处理？

**Comment by Windfarer** [comment, marked as answer]
get请求不支持这么玩，要么把请求参数字段改成扁平设计，要么用post

**Reply by xmsz** [reply]
ok

**Comment by xmsz** [comment]
有几个方向

1. 改为post，缺点就是改动了类型（下游和使用场景影响）。或者提供两个接口，高级搜索时再使用post。我更推荐
2. string to struct，前端提交的是filters_string={\"user_id\":1}，我们再通过json.转成struct，缺点就是类型变得无效，而且这种方式也不推荐
3. 扁平，第一种将user_id直接提到一层，第二种是变成filter.user_id，缺点如果参数太多(对大家)就很麻烦
4.

**Comment by daemon365** [comment]
在 HTTP 协议中GET请求的定义就是 “获得资源” 所以get的参数本就应该简单 扁平化

**Reply by xmsz** [reply]
没错 
因为我们大部分的请求列表接口都很简单，可能只有page和page_size这种，所以用的是Get
然后有一部分列表接口会有筛选的功能，就出现了filters={"user_id":1,"a":"b"}这种，为了不产生新的接口(避免其他问题)，所以用的还是Get的接口

然后，我们用过"filters[a]=b"和"filters={\"a\":\"b\"}"两种方式，前面这个proto不好定义，后面这个不优雅。
